packages/type-safe-api/src/construct/integrations/s3.ts

```diff
@@ -0,0 +1,179 @@
+/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0 */
+import { FeatureFlags, Stack } from "aws-cdk-lib";
+import {
+  CompositePrincipal,
+  Grant,
+  IRole,
+  Role,
+  ServicePrincipal,
+} from "aws-cdk-lib/aws-iam";
+import { IBucket } from "aws-cdk-lib/aws-s3";
+import * as perms from "aws-cdk-lib/aws-s3/lib/perms";
+import * as cxapi from "aws-cdk-lib/cx-api";
+import { ErrorIntegrationResponse } from "./error-integration-response";
+import { ErrorIntegrationResponses } from "./error-integration-responses";
+import {
+  ApiGatewayIntegration,
+  Integration,
+  IntegrationGrantProps,
+  IntegrationRenderProps,
+} from "./integration";
+import { generateCorsResponseParameters } from "../prepare-spec-event-handler/prepare-spec";
+import { Method } from "../spec";
+import { bucketInvocationUri } from "../spec/utils";
+
+export interface S3IntegrationProps {
+  /**
+   * The S3 bucket to be invoked on integration
+   */
+  readonly bucket: IBucket;
+
+  /**
+   * The IAM role to be used to grant permissions to the API Gateway to invoke the S3 bucket
+   * @default - a new role will be created
+   */
+  readonly role?: IRole;
+
+  /**
+   * The HTTP method to use when invoking the S3 bucket
+   * @default - integration method is used
+   */
+  readonly method?: Method;
+
+  /**
+   * The path override to use when invoking the S3 bucket
+   * @default - integration path is used
+   */
+  readonly path?: string;
+
+  /**
+   * The error integration response to use when the S3 bucket returns an error
+   * @default ErrorIntegrationResponses.catchAll()
+   */
+  readonly errorIntegrationResponse?: ErrorIntegrationResponse;
+}
+
+/**
+ * An S3 integration
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html
+ */
+export class S3Integration extends Integration {
+  private readonly bucket: IBucket;
+  private readonly role: IRole;
+  private readonly method?: Method;
+  private readonly path?: string;
+  private readonly errorIntegrationResponse?: ErrorIntegrationResponse;
+
+  constructor(props: S3IntegrationProps) {
+    super();
+
+    this.bucket = props.bucket;
+
+    if (props.role) {
+      this.role = props.role;
+    } else {
+      this.role = (this.bucket.node.tryFindChild("ExecutionRole") ??
+        new Role(this.bucket, "ExecutionRole", {
+          assumedBy: new CompositePrincipal(),
+        })) as IRole;
+    }
```

Comment:
I think better to create the role by default rather than attempting to reuse one already from the bucket construct - if users are worried about creating too many roles they can pass one in :)

Are you trying to "lazily" allow apigw to assume the role only for a specific method and path? If so I guess the empty `CompositePrincipal` probably works, will just need to update the `grant` to a `grantAssumeRole` I think :)

Reply:
But if two or more `S3Integration` are declared for the same bucket, we will have a conflict with roles right?

And yes in order to scope as down as possible I was thinking of declaring an empty role at init (just to store a reference and to get a roleArn in render) and then properly declare a `ServicePrincipal` grant per resource

But maybe it's over-engineering, I don't know, this is why I asked for feedback on this one

Reply:
Ahh I'd missed that - that's a very good point! :) I think the thing I found strange was that we're looking for the role in the bucket scope, maybe the simpler approach would be to move everything into the `render` method?

ie the render method could:

* Find or create the role on `props.scope` which is the same scope as the API
* Just create it with the apigw service principal (and if you want scope it down to the api but don't worry about scoping the source arn to the method/path as I think we might fill up the max trust policy size fairly quickly when reusing the role)
* Grant the permissions to the role based on the method (we can't do this in the `grant` method any more if the role is created in the `render` method)
* Then do the rest of what it already does :)

What do you think?

Reply:
Looks great!

Reply:
It's pushed :)

I'll improve tests, docs and examples later.

Where should I put examples ? I didn't find any with custom grant for lambda

Reply:
Awesome! Your latest changes look great! Thanks for all your hard work on this! 😄

So the `grant` method just gets called internally by the `TypeSafeRestApi` construct - the idea is the method is just a "hook" for the integration to grant any permissions it needs :)

I mentioned above that I thought we couldn't grant the permissions in the `grant` method any more... but what you've done is actually way cleaner, as it doesn't matter if `render` or `grant` is called first - it just uses the role if it's there or creates one - so great stuff there! 😄

Reply:
So you probably don't need to talk about any of the grant stuff in the docs :)